### ISSUE TITLE
How to make MOSA more accessible?

### ISSUE THREAD

**Issue by tgiphil** [mosa/MOSA-Project#149]
I would like to gather ideas on how to make MOSA more accessible to users. 

To kick start the discussion, here's my list of ideas:
- **Windows Installer** - installs the MOSA assemblies and pre-compiled tools.
- **Starter Templates integration with Visual Studio** - adds C# starter templates to Visual Studio under new projects.
- **Custom Project integration Visual Studio Integration** - adds the MOSA project type to Visual Studio to integrate the compiler and launcher.

This would be in addition to documentation and tutorials. 

Any other ideas?

**Comment by kiootic** [comment]
I think that these are not the primary reason of MOSA being inaccessible to end users; Rather, the primary reason is that the focus of MOSA project is different from what end users expect.

Let's take another popular managed OS toolkit, Cosmos, as an example. When end users go to their home page. They can immediately see how to make their own OS, with just one or two lines of line. Since most users want to build their own OS, they would go with using Cosmos.

However if they go to the home page of MOSA project, all they see is just a bunch of technical terms, vague description, and a complicated launcher UI. This would easily scare potential users if they don't understand what they mean.

Also, if my experience with Cosmos is still correct, they are mostly using a 'get it work' approach. i.e. Get the basic working, and improve later. IIRC, their compiler is a basic per-IL instruction translator, while MOSA has a complicated compiler pipeline with many optimization. Thus, although their output maybe inefficient, it works quite nicely without much catches. Therefore a great community is built with different libraries and tools.

For MOSA, due to the complexity of the project, many issues are about the compiler internal. Therefore, even if a user wants to participate in it, they may not have the required knowledge to contribute.

In conclusion, I think that the current situation is that, MOSA project is too technical and complex for average user to contribute or participate.

To make MOSA more accessible to average users, I suggest making the documentation less technical. For example, moving the 'Current Status' section from home page to wiki, adding sample codes to home page, and most importantly, actively promote the project. For example, creating a proper Wikipedia entry for MOSA, write some article about using MOSA, etc.

**Comment by charsleysa** [comment]
@tgiphil @kiootic COSMOS and MOSA are just too different to compare. 

Cosmos takes a let's just get something working so we can show people something approach.

Also they are heavily integrated into Visual Studio and from my opinion, is very far from being a viable operating system. 

In my opinion, I find the community around cosmos to be mostly comprised of people who like to install their userkit, compile it, and say they made an OS. Very few members of their community are involved in actual development of cosmos. 

MOSA differs greatly in that we aren't providing a plug and play solution for any hobby programmer to come along, compile the solution, and say they made an operating system. 

MOSA is laying the groundwork for developing a fully managed operating system. The CoolWorld, drivers, etc, are all just basic implementations to showcase the potential abilities of the MOSA Project. 

While having users would be nice, I would place emphasis on having developers that contribute to the development of the project but also branch out and create their own operating systems from the ground up using the tools provided by MOSA. I would not like to see MOSA turn into a OS-on-a-plate.

In response to this issue I would say that releasing a toolkit in which some of the tools, such as Explorer and Tiny Simulator, come precompiled could possibly be beneficial but the main components of the operating system itself such as the Runtime, the kernel, etc, should remain distributed as source code with CoolWorld being the base example as to what requirements there are from the pre-compiled compiler, and if anyone wants to change those requirements then they can modify the compiler to suit their needs and use their version instead of the precompiled version. 

This can lead to having two repositories, one with the tools and compiler, and one with the basic implementations (i.e. CoolWorld). 

A deciding factor needs to be what parts of the existing solution will be included in a precompiled version, and which parts will be distributed as source code showing the basic implementations.

**Comment by Hetal728** [comment]
When I first started trying to create a "demo" operating system with MOSA, one thing I got stumped on was trying to make it into a terminal environment. It was a struggle trying to get the "backspace", "enter", etc keys to work. I think I got it working though, not sure (it's been a while). Maybe make CoolWorld more "advanced" but still basic?

**Comment by Rew** [comment]
I've been debating following up on this for awhile, and my compiler has me really annoyed today.  So, I'll give you my perspective.

I started OS dev a few years ago.  I ran into the first barrier that everyone encounters, getting anything to compile in the "right" way.  Setting up an environment can just be creating a cross compiler (for some people), or it can be creating a virtual machine to install linux, learning eclipse, and learning the gcc toolchain.  I was one of the later.  This process alone can demand substantial endurance to get through.

After getting a bit past the standard x86 barebones OS using c/asm, I ended up frustrated with the development environment.  I currently work in C# all day everyday and I find the development experience of VisualStudio with the productivity capabilities allowed by "managed" code to be something that is annoying to live without.  

I started with the idea of a Visual Studio addin for a C/asm environment.  Sadly this still requires a learning curve to get a cross compiler running (hint: you can't use the micrsoft toolchain, go learn cygwin and gcc!).  After I got it working, I was still annoyed that the unmanaged environment didn't get me the development experience I wanted.  So I started investigating the possibility of an OS written in C#.  At the time, there was far less information available on this front.  I was not initially aware of MOSA, TysOS, or any of the other similar projects out there.  All I could find at the time was Cosmos.  For me, I was really turned off by Cosmos.  I didn't want the OS development work done for me.  I didn't like Console.Write just printing to the screen, it hides the purpose of OS development for me.  I wanted to still write every component of the OS and have full control of everything the OS does.  

So... a few years later I've written (and rewritten 10 times over) a Lexer, Parser, optimizer, Assembler, Linker, etc in a long story that isn't relevant here.  Basically, I spent several years on my C# toolkit development to get the developer "experience" I enjoy for OS development. (My OS has not progressed)

I think from the outside Mosa as it stands feels like another toolkit to learn and it feel kludgy to use for development.  

I think tgiphil's first post hit what is needed accurately.  To get wider usage, Mosa needs to cater to the experience that managed developers are accustomed to.  Or, another tool that uses MOSA as the backbone will need to cater to it.

**Comment by charsleysa** [comment]
@rew the problem is that Mosa is a toolkit but not the same kind of toolkit like COSMOS where everything is done for you. 

Mosa is a toolkit where it has components and reference implementations ready to use but you are the one to develop the OS how ever you want and it doesn't hide how everything works. 

The MOSA Project itself should not cater to a single development environment but I would support the idea of having Visual Studio plugins and such. 

I think one of the issues stopping mosa becoming accessible is one that is unavoidable, and that is knowing how the CLR works at a low level.

I think lots of people are enticed at the thought of making an OS with C# but don't have the necessary understanding of C# and CLR so they find it difficult.

I was one of those people and it took me roughly two years before I could jump into mosa and even then I had a lot to learn before I could understand how it all works. Even today there are parts of the compiler (mainly the optimizers) that I don't understand properly.

**Comment by Arakis** [comment]
Before writing VS plugins, it would be better to have some non-plugin tools like an GDB debugger interface ( http://www.gnu.org/software/gdb/ ). The GDB debugger is runable on windows, too, on the other side, the microsoft visual studio debugger is not runable on unix ("of course"). You should not do the same mistake than cosmos and go into an oneway street, and force the users to use microsoft windows / visual studio. Than the opposite will happen: Mosa will become less accessible (because linux community will be effectively excluded). When the people see, that only VS plugins will lovely maintenanced, but non VS extensions (like GDB) will not supported, linux users could feel excluded. So a big part of the possible community is not "accessible". By the way: C# Development on linux exists...it not only the domain of microsoft windows.

Maybe we should not even use GDB, maybe we using an own debugger. One matter, why you "should" not write an Visual studio plugin is, too, that you have it to maintain from VS version to VS version. And when you write an own GUI: Please do not use the third party controls. they only looks cool, but do not work on mono/unix.

**Comment by tgiphil** [comment]
@Rew  Thanks for the comments. VS integration has started - but it also a steep time commitment and learning curve. I hope someone will volunteer to jump start it. ;)

I've love to see that Lexer thru to Linker chain of tools. Any chance you can open source?

**Comment by Rew** [comment]
@charsleysa I'm aware of what MOSA is now, and I believe it would be a great tool.  However, I guess my background kind of proves your point.  I have spent years doing what MOSA does and have a great understanding of CLI, IL compilers, and the (so far) necessary hooks between the compiler and the OS.    I guess it is easy to take for granted what you have learned.  I ended up with a similar approach to MOSA in using attributes to identify necessary OS features to the compiler.  Some thought would really have to be put into reducing the learning curve necessary to get at least a basic OS running, even if a developer would have to learn a bit more about the CLR and BCL to get it working in the end.

@Arakis I'm with you on the basic idea, but it's difficult to separate the debugger and even the compiler GUI from the development environment.  People that would typically be attracted to the idea of MOSA are looking for what I called the "developer experience"  that an IDE for a managed language can provide, both at design time and run-time.  While Visual Studio would be an easy target as a starting point, it doesn't mean that it is the only option.  Also, adding visual studio support doesn't mean it is too challenging to add support for another IDE in the future.  After writing language extensions, custom debugging, and custom project types, I do not believe that it would be that difficult to keep the majority of code outside of the visual studio layer.  You could have a debugging environment that is platform agnostic and just have a small "glue" layer to finish off the experience.  This is basically what Roslyn ended up as.  Yes, Roslyn ended up as a key component to the Visual Studio experience, but it can also be used separately.

@tgiphil Sadly the full chain no longer exists.  After doing my first couple of OS builds with an entire toolchain written by me I started working to get beyond C# 1 and 2.  Somewhere in generics and lambda expressions I realized that I could spend a full time job trying to keep a C# parser up with the changing language spec.  Additionally, the performance optimizations that I ended up spending significant time on were not getting me anywhere towards a goal of writing an OS. At that point I started using the Roslyn preview for lex/parse and still using the AST to convert directly to my intermediate representation.  Again after getting further in the language specification I realized the scope of the problem and regressed to using pre-compiled IL as my input.  Now I have a compiler that is very similar to MOSA, but is a couple of years behind in terms of performance optimizations.

I'm actually at a crossroads where I am trying to decide if pursuing the compiler is really keeping me entertained anymore or if I would like to let tgiphil do the awesome compiler work that he's done and focus more on the toolchain that uses that compiler.  I was sort of waiting to see the direction that you guys decided to go and make a decision if I would be able to help the project or not.
